Scrum 168 노트 검색 화면 구현

inwoodev opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [x] 버그 수정
- [x] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [ ] Other... 추가 설명 요망:


## 배경
- 노래 검색 및 관련 노래에 대한 노트 갯수를 리스트화 하여 보여주는 화면이 필요합니다.
- 특정 노래에 대한 노트 검색 결과를 보여주는 화면이 필요합니다.
- 노트의 상세화면 및 관련 댓글을 보여주는 화면이 필요합니다.
- 특정 노트에 댓글을 작성할 수 있는 기능이 필요합니다.

## 목표

- [x] 노래 검색 및 관련 노래에 대한 노트 갯수를 보여주는 화면 & 로직 구현
  - [x] 노래를 실시간 검색할 수 있는 기능 구현
  - [x]  키보드 높이에 맞게 collectionView인셋을 조정하여 모든 검색결과를 키보드가 뜬 상태에서 확인할 수 있도록 기능 구현
- [x] 특정 노래에 대한 노트 검색 결과를 보여주는 화면 & 로직 구현
  - [x] 가사 있는 노트만을 보여주는 기능 구현 
  - [x] 노트 좋아요, 북마크, 삭제 기능 구현 
- [x] 노트의 상세화면 및 관련 댓글을 보여주는 화면 & 로직 구현
  - [x] 댓글 작성, 삭제 기능 구현
  - [x] 댓글 작성시 TextView크기 변경에 따른 UI변경 기능 구현

## 결과 (Optional)

### 노래 & 노트갯수 검색 화면
|기본 검색 | 실시간 검색 반영|
|:-----:|:-----:|
| ![GIF 1](https://github.com/user-attachments/assets/d51789ad-7f10-46f8-8e03-90c089ff75c8) | ![GIF 2](https://github.com/user-attachments/assets/42773dd3-ffec-4289-a7f3-46b584effae1) |

### 노트 상세 화면
|노트 상세 화면 (실제 데이터) | 노트 상세화면 (테스트 데이터)|
|:-----:|:-----:|
| ![GIF 1](https://github.com/user-attachments/assets/ff40a218-6d1e-427a-80f7-e5178f19771a) | ![GIF 2](https://github.com/user-attachments/assets/f504781c-e29e-4383-8eed-da495ce26556) |

### 노트 상세 & 댓글 화면
|노트 & 댓글 화면 | 댓글 작성 중 화면|
|:-----:|:-----:|
| ![GIF 1](https://github.com/user-attachments/assets/5aa7bf87-3a84-4fac-aa6d-a3a78c3b00e2) | ![GIF 2](https://github.com/user-attachments/assets/6e046268-ff57-428a-a374-61e349216e1c) |

inwoodev committed:
feat: SCRUM-169 노트 실시간 검색 화면 및 로직 구현

inwoodev committed:
refactor : mainCollectionView → homeCollectionView로 네이밍 수정

inwoodev committed:
refactor : 노트메뉴, 노트 음악 실행 부분을 컴포넌트화 하여 재사용성 ↑

- 노트 검색 상세화면 VC 그리고 노트 및 댓글화면 VC에서도 동일한 기능이 필요로 한 상황에서
프로토콜 기본구현을 활용하여 노트 메뉴 그리고 노트 음악 실행 부분을 컴포넌트화 하여 중복 코드 방지

inwoodev committed:
add: 노트상세화면 및 노트 댓글화면에 필요한 이미지에셋 추가

목록:
- checkmark active, inactive
- replay active, inactive

inwoodev committed:
feat / SCRUM-170: 노래에 대한 노트 검색 결과 화면 및 로직 구현

- 가사 포함된 노트만 보기 텝할 경우 관련 api호출 및 노트 데이터 업데이트 로직 구현
- 노트 삭제, 좋아요, 북마크 기능 추가

inwoodev committed:
refactor: homeViewController collectionView 데이터 처리방식을 diffable datasource & snapshot 방식으로 변경

inwoodev committed:
refactor : NoteDetailCollectionView의 snapshot 버그 수정

- snapshot처리 방식 로직을 다음어서 좋아요 또는 북마크 할 때 마다 cell이 번쩍이던 애니메이션이 없어지도록 수정

inwoodev committed:
refactor: 키보드 높이에 따라 NoteSearchTableView의 bottom inset을 조정하는 로직 추가

inwoodev committed:
refactor : HomeViewController의 diffableDataSource적용 중 누락된 부분 추가

- UICollectionViewDataSource제거
- 각 섹션을 snapshot을 활용하여 업데이트하도록 로직 수정
- supplementaryViewProvider를 활용하여 UICollectionReusableView를 관리하도록 로직 수정

inwoodev committed:
refactor, feat: 노트와 댓글 관련 api 추가 & 그외 경로 수정

- 노트& 댓글을 가져오는 api 추가
- 노트 삭제, 수정 경로 수정

inwoodev committed:
feat: 노트 댓글 불러오기, 댓글 삭제, 그리고 댓글 작성 로직 추가

- 관련 dto & Model 추가
- 관련 Service & UseCase 추가

inwoodev committed:
feat: 댓글 작성 UI 컴포넌트 구현

- maxNumberOfLines에 도달하기 전까지 textView가 상하로 확장되며
최대 maxNumberOfLines를 넘게 될 경우 스크롤이 되는 textView구현

- 내용 유무에 따라서 전송 버튼이 enable/disable되는 기능 구현
- textView 크기 변경에 따라 전체 뷰의 크기가 조정되도록 UI 구현
- 전송 버튼을 눌렀을 때 이벤트를 방출하는 기능 구현
- WriteCommentView의 높이가 변경될 때 이벤트를 방출하는 기능 구현

inwoodev committed:
feat: 댓글&작성 화면 구현

- 노트셀, 댓글셀을 다른 섹션으로 갖고 있는 컬렉션 뷰 구현
- 화면 키보드 유무에 따라 댓글 작성UI의 위치가 변경되도록 로직 구성
- 댓글작성 UI 크기 변경에 따라서 전체 UI 위치를 조정하는 로직 구현
- 댓글작성 textView 외 다른 UI 터치시 키보드를 dismiss하는 기능 구현

inwoodev committed:
feat/SCRUM-171: 노트 상세정보 및 댓글 UI & 로직 구현

- noteID를 활용하여 노트 단건 & 관련 댓글을 화면에 보여주는 기능 추가
- 노트 단건 좋아요, 북마크, 삭제 기능 추가
- 댓글 작성 기능 추가
- 댓글 삭제 기능 추가
- 댓글 작성, 삭제시 댓글을 최신화 하는 기능 추가
- pull-to-refresh기능을 활용하여 노트 및 댓글을 최신화 하는 기능 추가

inwoodev committed:
refactor: VC 구분 주석 수정

inwoodev committed:
refactor: 댓글 컬랙션뷰의 수직 scroll indicator 숨김 처리

inwoodev commented on Projects/Core/Network/Interface/Sources/DTO/SearchedNoteResponse.swift:
```diff
@@ -0,0 +1,86 @@
+        case artistName = "name"
+    }
+    
+    // Nested json을 파싱하기 위한 커스텀 이니셜라이저
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.id = try container.decode(Int.self, forKey: .id)
+        self.songName = try container.decode(
+            String.self,
+            forKey: .songName
+        )
+        self.imageUrl = try container.decode(
+            String.self,
+            forKey: .imageUrl
+        )
+        self.noteCount = try container.decode(
+            Int.self,
+            forKey: .noteCount
+        )
+        
+        let artistContainer = try container.nestedContainer(
+            keyedBy: ArtistKeys.self,
+            forKey: .artist
+        )
+        self.artistName = try artistContainer.decode(
+            String.self,
+            forKey: .artistName
+        )
+    }
```
서버에서 내려주는 데이터에서 노트 검색화면에 있어서 불필요한 부분이 많아서 부득이하게 codingkey를 활용하여 필요한 데이터만 파싱하도록 하였습니다.

inwoodev commented on Projects/Feature/Main/Interface/Sources/Protocols/NoteMenuHandling.swift:
```diff
@@ -0,0 +1,45 @@
+import Domain
+import Shared
+
+public protocol NoteMenuHandling where Self: UIViewController {
+    var userInfo: UserInformation? { get }
+    var onReportNote: PassthroughSubject<Int, Never> { get }
+    var onEditNote: PassthroughSubject<Int, Never> { get }
+    var onDeleteNote: PassthroughSubject<Int, Never> { get }
+    
+    func makeNoteMenuViewController(checking note: Note) -> NoteMenuViewConroller?
+}
+
+public extension NoteMenuHandling {
+    func makeNoteMenuViewController(checking note: Note) -> NoteMenuViewConroller? {
+        if let userId = self.userInfo?.userID {
+            let bottomSheetHeight: CGFloat = userId == note.publisher.id ? 180 : 130
+            
+            let menuType = userId == note.publisher.id ? NoteMenuType.me : NoteMenuType.other
+            
+            let noteMenuViewController = NoteMenuViewConroller(
+                noteID: note.id,
+                bottomSheetHeight: bottomSheetHeight,
+                bottomSheetView: NoteMenuView(menuType: menuType),
+                onReport: self.onReportNote,
+                onEdit: self.onEditNote,
+                onDelete: self.onDeleteNote
+            )
+            noteMenuViewController.modalPresentationStyle = .overFullScreen
+            
+            return noteMenuViewController
+        }
+        
+        return nil
+    }
+}
```
홈화면, 그리고 노트 검색결과화면, 커뮤니티 화면 등등 노트 신고, 수정, 삭제에 있어서 동일한 bottomSheetViewController를 필요로 합니다. 이 부분은 추후 변경된다 하더라도 공통적으로 변경될 것이라 판단하여 프로토콜 & 기본 구현을 활용하여 위와 같이 컴포넌트화 하였습니다. 이를 통해 중복 코드작성을 조금이나마 줄일 수 있습니다!

inwoodev commented on Projects/Feature/Main/Interface/Sources/Protocols/NoteMusicHandling.swift:
```diff
@@ -0,0 +1,46 @@
+    func openYouTube(query: String)
+}
+
+public extension NoteMusicHandling {
+    var youTubeMusicURLScheme: String {
+        return "youtubemusic://"
+    }
+    
+    func isYouTubeMusicInstalled() -> Bool {
+        if let url = URL(string: youTubeMusicURLScheme) {
+            return UIApplication.shared.canOpenURL(url)
+        }
+        return false
+    }
+    
+    func openYouTube(query: String) {
+        let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)
+        
+        if isYouTubeMusicInstalled() {
+            let youtubeMusicPath = "\(youTubeMusicURLScheme)search/\(encodedQuery ?? query)"
+            
+            if let url = URL(string: youtubeMusicPath) {
+                UIApplication.shared.open(url)
+            }
+        } else {
+            let youtubeMusicWebPath = "https://music.youtube.com/search?q=\(encodedQuery ?? query)"
+            
+            if let url = URL(string: youtubeMusicWebPath) {
+                UIApplication.shared.open(url)
+            }
+        }
+    }
+}
```
NoteMenuHandling과 비슷한 맥락으로 컴포넌트화 진행하였습니다!

inwoodev commented on Projects/Feature/Main/Interface/Sources/ViewControllers/HomeViewController.swift:
```diff
@@ -106,47 +188,99 @@ public class HomeViewController: UIViewController, NoteMenuHandling, NoteMusicHa
 
+            homeDataSource.apply(snapshot, animatingDifferences: false)
+        }
+    }
```
노트 추가 및 삭제시 애니메이션이 들어가는 것은 자연스러우나 좋아요 및 북마크시 애니메이션이 들어가는 것은 부자연스럽다고 판단하였고 snapshot의 갯수 차이 유무에 따라서 snapshot업데이트에 애니메이션을 주거나 주지 않는 방향으로 노트 컬렉션 뷰 업데이트를 구현하였습니다.

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/WriteCommentView.swift:
```diff
@@ -0,0 +1,242 @@
+        maxTextViewHeight = (lineHeight * CGFloat(maxNumberOfLines)) + textInsets.top + textInsets.bottom
+    }
+    
+    // UITextView의 내용이 변경될 때 호출
+    func textViewDidChange(_ textView: UITextView) {
+        // 버튼 활성화 상태 업데이트
+        self.sendCommentButton.isEnabled = !textView.text.isEmpty
+        
+        // Placeholder 레이블을 숨기거나 표시
+        placeholderLabel.isHidden = !textView.text.isEmpty
+        
+        // 텍스트 크기 계산
+        let size = textView.sizeThatFits(
+            CGSize(width: textView.frame.width,
+            height: CGFloat.greatestFiniteMagnitude)
+        )
+        
+        // 동적 높이 조정
+        if size.height <= maxTextViewHeight {
+            textView.isScrollEnabled = false
+            textView.flex.height(size.height).markDirty() // FlexLayout이 크기를 다시 계산하도록 표시
+            
+            let changedContainerHeight: CGFloat = size.height + 40
+            self.didChangeHeightSubject.send(changedContainerHeight)
+            flexContainer.flex.height(changedContainerHeight).markDirty()
+            
+        } else {
+            textView.isScrollEnabled = true
+            textView.flex.height(maxTextViewHeight).markDirty() // 최대 높이로 제한
+            self.didChangeHeightSubject.send(maxTextViewHeight + 40)
+            flexContainer.flex.height(maxTextViewHeight + 40).markDirty()
+        }
+        
+        // 레이아웃을 다시 정리
+        flexContainer.flex.layout(mode: .adjustHeight)
+    }
```
텍스트뷰 의 댓글 높이에 따라서 textView의 height 및 flexContainer의 높이를 조정하는 부분입니다.

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/WriteCommentView.swift:
```diff
@@ -0,0 +1,242 @@
+        )
+    }
+    
+    @objc private func sendComment() {
+        self.didTapSendSubject.send(self.textView.text)
+        
+        self.textView.text = ""
+        
+        self.textView.resignFirstResponder()  // 키보드 닫기
+        self.textView.flex.height(originalTextViewHeight).markDirty()
+        self.didChangeHeightSubject.send(originalTextViewHeight + 40)
+        
+        self.sendCommentButton.isEnabled = false
+        
+        // 레이아웃 다시 계산
+        self.flexContainer.flex.height(80).markDirty()
+        self.flexContainer.flex.layout(mode: .adjustHeight)
+    }
```
댓글 전송 후 후처리를 하는 부분입니다.

1. textView의 텍스트를 초기화 하며
2. 키보드를 숨기고
3. textView의 크기를 최초 설정으로 되돌리며
4. 전송 버튼을 비활성화 하며
5. 마지막으로 flexContainer의 높이를 재조정하게 됩니다.

inwoodev commented on Projects/Feature/Main/Interface/Sources/Views/NoteCommentsView.swift:
```diff
@@ -0,0 +1,210 @@
+    
+    // MARK: - Binding
+    
+    private func bindHeight() {
+        CombineKeyboard.keyboardHeightPublisher
+            .sink { [unowned self] keyboardHeight in
+                UIView.animate(withDuration: 0.3) {
+                    // 키보드 높이만큼 collectionview에 inset 부여
+                    self.noteCommentsCollectionView.contentInset = .init(
+                        top: 0,
+                        left: 0,
+                        bottom: keyboardHeight,
+                        right: 0
+                    )
+                    // 키보드 높이 변경시 writeComment의 위치 조정
+                    self.writeCommentView.flex.position(.relative).bottom(keyboardHeight)
+                    
+                    
+                    self.flexContainer.flex.layout(mode: .adjustHeight)
+                }
+            }
+            .store(in: &cancellables)
+        
+        self.writeCommentView.didChangeHeightPublisher
+            .sink { [unowned self] updatedHeight in
+                
+                // WriteCommentView의 높이가 바뀔 때 마다 flexContainer의 height를 재조정
+                self.flexContainer.flex.markDirty()
+                self.flexContainer.flex.layout(mode: .adjustHeight)
+            }
+            .store(in: &cancellables)
+    }
```
댓글 작성시 UI를 조정하는 핵심 로직은 이부분 입니다.
1. 키보드가 화면에 보여질 경우 writeComment뷰를 키보드뷰 하단에 위치 시키며
noteCommentsCollectionView의 인셋을 키보드 높이만큼 조정합니다.

2. 키보드 높이가 아닌 **writeCommentView**의 높이가 변경될 때 마다
NoteCommentsView의 크기 자체를 재조정하게 됩니다. 해당 부분이 빠지게 될 경우 writeComment뷰에서 줄넘김을 하여 WriteCommentView의 크기가 커지게 될 경우 키보드 밑으로 WriteCommentView가 커지게 되며 부분적으로 키보드가 댓글 작성 화면을 가리게 되는 문제가 발생하게 됩니다.

derrickkim0109 replied:
와..고생하셨습니다.

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/API/FeelinAPI.swift:
```diff
@@ -94,12 +99,34 @@ extension FeelinAPI: HTTPNetworking {
         case .postLikes(let noteID),
                 .deleteLikes(let noteID),
                 .postBookmarks(let noteID),
-                .deleteBookmarks(let noteID),
-                .deleteNote(let noteID):
+                .deleteBookmarks(let noteID):
```
혹시 정렬이 안된걸까요?

inwoodev replied:
오 저 같은 경우 case에서 줄넘김을 하게 되면 인의적으로 줄을 맞추지 않는 이상 위와 같이 정렬이 되더라구요.
정렬해볼게요~

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/DTO/DeleteCommentResponse.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+public struct DeleteCommentResponse: Decodable {
+    public let success: Bool
+}
```
제가 이렇게 넘어오는 Response `FeelinSuccessResponse` 이 객체를 만들어뒀는데 혹시 통일하는 건 어떻게 생각하세요?

inwoodev replied:
네네 저도 동일한 모델인걸 인지하였는데 delete, post comment 둘 다 추후 response모델이 변경이 있을지도 모르는 점을 염두 해 두고 나눠 봤는데요...음 일단 현재로서는 오버 엔지니어링인듯 해요 ㅎㅎ. `FeelinSuccessResponse`로 통일 하시죠~!

derrickkim0109 replied:
좋습니다~!

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/DTO/NoteCommentsResponse.swift:
```diff
@@ -0,0 +1,40 @@
+        case comments
+    }
+    
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        
+        self.note = try NoteResponse(from: decoder)
+        self.commentsCount = try container.decode(Int.self, forKey: .commentsCount)
+        self.comments = try container.decode(
+            [CommentResponse].self,
+            forKey: .comments
+        )
+    }
```
NoteCommentsResponse는 이렇게 decode를 안에서 진행하신 이유가 있을까요?

inwoodev replied:
저희 노트 단건 조회 json 구조를 보시면 아래와 같습니다:
```
{
  "id": 320,
  "content": "노트 내용",
  "status": "PUBLISHED",
  "createdAt": "2024-09-02 13:01:54",
  "lyrics": {
    "lyrics": "가사",
    "background": "DEFAULT"
  },
  "publisher": {
    "id": 317,
    "nickname": "nickname",
    "profileCharacterType": "poopHair"
  },
  "song": {
    "id": 319,
    "name": "Flying Bob",
    "imageUrl": "https://i.scdn.co/image/ab67616d0000b2739c3a4e471c5e82a457dce2c0",
    "artist": {
      "id": 318,
      "name": "검정치마",
      "imageUrl": "https://i.scdn.co/image/ab6761610000e5eb8609536d21beed6769d09d7f"
    }
  },
  "commentsCount": 1,
  "comments": [
    {
      "id": 321,
      "content": "content",
      "createdAt": "2024-09-02 13:01:54",
      "writer": {
        "id": 317,
        "nickname": "nickname",
        "profileCharacterType": "poopHair"
      }
    }
  ],
  "likesCount": 0,
  "isLiked": false,
  "isBookmarked": false
}
```

이걸 그대로 swift dto로 변환하게 되면 이렇게 됩니다!

``` swift
import Foundation

public struct NoteResponse: Decodable {
    public let id: Int
    public let content: String
    public let status: NoteStatusDTO
    public let createdAt: Date
    public let lyrics: LyricsDTO?
    public let publisher: UserDTO
    public let song: SongResponse
    public let commentsCount: Int
    public let likesCount: Int
    public let isLiked: Bool
    public let isBookmarked: Bool
    public let comments: [Comment]     // 기존 NoteResponse에 comment배열이 추가됩니다. 
}

```

그리고 만약 noteCount까지 여기에 추가된다면 NoteResponse모델은 더 복잡해지겠죠

``` swift
import Foundation

public struct NoteResponse: Decodable {
    public let id: Int
    public let content: String
    public let status: NoteStatusDTO
    public let createdAt: Date
    public let lyrics: LyricsDTO?
    public let publisher: UserDTO
    public let song: SongResponse
    public let commentsCount: Int
    public let likesCount: Int
    public let isLiked: Bool
    public let isBookmarked: Bool
    public let comments: [Comment]?     // 해당 프로퍼티가 아예 없는 경우 존재(ex: 홈화면)
    public let noteCount: Int?                    // 해당 프로퍼티가 아예 없는 경우 존재(ex: 홈화면)
}
```

이렇게 기존 모델에 불필요할 수 있는 프로퍼티들을 추가하는 방향으로 수정해 나간다면 사용성 측면에서 너무 안좋을 것 같다는 생각이 들었습니다. optional들을 다 처리해야 하는거죠.

그런 이유로 이렇게 noteResponse와 CommentResponse를 분리할 수 있는 dto를  만들었습니다:)

``` swift
import Foundation

public struct NoteCommentsResponse: Decodable {
    public let note: NoteResponse
    public let commentsCount: Int
    public let comments: [CommentResponse]
    
    private enum CodingKeys: String, CodingKey {
        case commentsCount
        case comments
    }
    
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        self.note = try NoteResponse(from: decoder)
        self.commentsCount = try container.decode(Int.self, forKey: .commentsCount)
        self.comments = try container.decode(
            [CommentResponse].self,
            forKey: .comments
        )
    }
    
    public init(
        note: NoteResponse,
        commentsCount: Int,
        comments: [CommentResponse]
    ) {
        self.note = note
        self.commentsCount = commentsCount
        self.comments = comments
    }
}
```

derrickkim0109 replied:
아하. 설명 감사합니다. 이해했습니다:)

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/DTO/PostCommentResponse.swift:
```diff
@@ -0,0 +1,12 @@
+
+import Foundation
+
+public struct PostCommentResponse: Decodable {
+    public let success: Bool
+}
```
이 Response도 하나로 통일해서 사용해도 되지 않을까요?

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/DTO/SearchedNoteResponse.swift:
```diff
@@ -0,0 +1,86 @@
+    public init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self.id = try container.decode(Int.self, forKey: .id)
+        self.songName = try container.decode(
+            String.self,
+            forKey: .songName
+        )
```
제가 init에서 try container.decode(_:)를 사용한 방식이 익숙하지 않아서.. 이 방식의 장점이 어떤 부분인지 설명해주실 수 있을까요?

inwoodev replied:
제가 위와 같은 방법으로 진행한 이유는 다음과 같습니다.

SearchedNoteViewController에서 활용될 모델을 구현하기 위해 서버쪽에서 내려주는 api를 살펴봤는데요. 요런 구조로 데이터를 내려주더라구요.

```
{
  "pageNumber": 0,
  "hasNext": true,
  "data": [
    {
      "id": 481,
      "name": "Flying Bob",
      "imageUrl": "https://i.scdn.co/image/ab67616d0000b2739c3a4e471c5e82a457dce2c0",
      "noteCount": 0,
      "artist": {
        "id": 482,
        "name": "검정치마",
        "imageUrl": "https://i.scdn.co/image/ab6761610000e5eb8609536d21beed6769d09d7f"
      }
    },
    {
      "id": 483,
      "name": "Flying Bob",
      "imageUrl": "https://i.scdn.co/image/ab67616d0000b2739c3a4e471c5e82a457dce2c0",
      "noteCount": 0,
      "artist": {
        "id": 484,
        "name": "검정치마",
        "imageUrl": "https://i.scdn.co/image/ab6761610000e5eb8609536d21beed6769d09d7f"
      }
    }
  ]
}
```
여기에서 data부분을 살펴보시면 저희가 기존에 만들어 둔 NoteResponse를 사용할 수도 있겠다 생각이 드실 수도 있어요.

``` swift
public struct NoteResponse: Decodable {
    public let id: Int
    public let content: String
    public let status: NoteStatusDTO
    public let createdAt: Date
    public let lyrics: LyricsDTO?
    public let publisher: UserDTO
    public let song: SongResponse
    public let commentsCount: Int
    public let likesCount: Int
    public let isLiked: Bool
    public let isBookmarked: Bool
}
```

하지만 여기서 함정이 있는게 기존 구현된 NoteResponse에는 noteCount라는 변수가 없습니다. 따라서 기존에 만들어둔 NoteResponse를 사용한다면 NoteResponse 구조를 수정해야 하거나 아니면 매우 유사한 Response를 별도로 구현해야 합니다.(noteCount를 아래처럼 optional하게 둘 수도 있겠죠:)


``` swift
public struct NoteResponse: Decodable {
    public let id: Int
    public let content: String
    public let status: NoteStatusDTO
    public let createdAt: Date
    public let lyrics: LyricsDTO?
    public let publisher: UserDTO
    public let song: SongResponse
    public let commentsCount: Int
    public let likesCount: Int
    public let isLiked: Bool
    public let isBookmarked: Bool
    public let noteCount: Int?    // 수정 부분
}
```

그런데 이렇게 옵셔널하게 두게 해당 NoteResponse를 활용할 때 optional체크를 해야한다는 불편함이 예상되었어요.

그래서 그냥 SearchedNoteViewController에 특화된 dto로 아예 파싱과정에서 변환을 하는게 편하다고 판단 되어서  CodingKeys를 활용해서 커스텀하게 만들어 봤습니다.

장점이라고 한다면 받는 쪽에서 아예 사용하기 편한 구조로 파싱이 되기 때문에 dto에서 모델로 변환할 때 복잡도가 내려간다는 부분이 있을듯 싶습니다.

``` swift
import Foundation

public struct SearchedNotesResponse: Decodable {
    public let pageNumber: Int
    public let hasNext: Bool
    public let data: [SearchedNoteResponse]
    
    public init(
        pageNumber: Int,
        hasNext: Bool,
        data: [SearchedNoteResponse]
    ) {
        self.pageNumber = pageNumber
        self.hasNext = hasNext
        self.data = data
    }
}

public struct SearchedNoteResponse: Decodable {
    public let id: Int
    public let songName: String
    public let artistName: String
    public let imageUrl: String
    public let noteCount: Int
.
.
.
}
```

다만 단점이라면 decoder를 활용해서 파싱하는 부분이 비교적 복잡하네요... NoteResponse와 비슷한 모델을 하나 더 만들고 SearchedNote로 변환할 때 사용하기 편한 구조로 만드는 방법도 있겠지만...뭔가 프로퍼티들이 다 사용되지 않는 dto를 만드는데에 거부감이 있어서 위와 같이 구현하였습니다 :)

derrickkim0109 commented on Projects/Core/Network/Interface/Sources/API/FeelinAPI.swift:
```diff
@@ -26,6 +26,11 @@ public enum FeelinAPI<R> {
     case deleteNote(noteID: Int)
+    case getSearchedNotes(pageNumber: Int, pageSize: Int, query: String)
+    case getSongNotes(cursor: Int?, size: Int, hasLyrics: Bool, songID: Int)
+    case getNoteWithComments(noteId: Int)
```
noteID로 통일해주실 수 있을까요?

inwoodev replied:
넹~~

derrickkim0109 commented on Projects/DependencyInjection/Sources/DIContainer/Inject/InjectIdentifier.swift:
```diff
@@ -69,14 +69,18 @@ public extension InjectIdentifier {
     
-    static var artistPaginationService: InjectIdentifier<ArtistPaginationServiceInterface> {
-        .by(type: ArtistPaginationServiceInterface.self, key: "artistPaginationService")
+    static var artistPaginationService: InjectIdentifier<KeywordPaginationServiceInterface> {
+        .by(type: KeywordPaginationServiceInterface.self, key: "artistPaginationService")
     }
```
artistPaginationService만 key와 type의 이름이 다른데 다른 이유가 있으실까요?

inwoodev replied:
```
   static var artistPaginationService: InjectIdentifier<KeywordPaginationServiceInterface> {
        .by(type: KeywordPaginationServiceInterface.self, key: "artistPaginationService")
    }
    
    static var searchedNotePaginationService: InjectIdentifier<KeywordPaginationServiceInterface> {
        .by(type: KeywordPaginationServiceInterface.self, key: "searchedNotePaginationService")
    }
```

제가 searchedNotePaginationService를 추가하는걸 빼먹었네요. 두 객체 모두 KeywordPaginationServiceInterface를 채택하게 설계하였습니다!!

derrickkim0109 commented on Projects/Domain/Note/Interface/Sources/Services/CommentAPIService.swift:
```diff
@@ -0,0 +1,49 @@
+public protocol CommentAPIServiceInterface {
+    func getNoteWithComments(noteID: Int) -> AnyPublisher<NoteCommentsResponse, NoteError>
+    func postComment(body: PostCommentRequest) -> AnyPublisher<PostCommentResponse, NoteError>
+    func deleteComment(commentId: Int) -> AnyPublisher<DeleteCommentResponse, NoteError>
```
commentId 도 통일되면 좋을것 같습니다.

inwoodev replied:
넵!!

derrickkim0109 commented on Projects/Domain/Note/Interface/Sources/Services/SearchedNotePaginationService.swift:
```diff
@@ -0,0 +1,50 @@
+
+import DomainSharedInterface
+
+final public class SearchedNotePaginationService: KeywordPaginationServiceInterface {
```
혹시 SearchedNotePaginationService 이 객체는 생성 때마다 초기화할 필요는 없을까요? 
저는 곡 검색에서 뒤로가기하고 다시 ViewModel을 생성할 때, 초기화 시켜주지 않으면 기존 값들이 남아있는 문제가 있었어요.

inwoodev replied:
오 네 분명 pop을 해서 다시 해당 화면으로 돌아오게 되면 기존 값들이 남아있을거에요. 다만 이게 문제가 되지 않을듯 합니다.

곡 검색 결과를 활용하여 컬렉션 뷰 외에 추가적인 바인딩 처리가 되어 있지 않기 때문에 사이드 이펙트는 없을 듯 합니다. 다만 이건 화면들을 합쳤을 때 한 번 더 체크 해 봐도 좋을 듯 합니다 :)

derrickkim0109 replied:
네 알겠습니다~!

derrickkim0109 commented on Projects/Feature/Main/Example/Sources/SceneDelegate.swift:
```diff
@@ -62,67 +62,190 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
-            setBookmarkUseCase: SetBookmarkUseCase(noteAPIService: noteAPiService),
-            deleteNoteUseCase: DeleteNoteUseCase(noteAPIService: noteAPiService)
-        )
+//        DIContainer.standard.register(.networkProvider) { _ in
```
여기는 저도 많이 건들여서 컴플릭이 많이 날 것 같네요..

inwoodev replied:
잡느라 애좀 먹었습니다 ㅠㅠ. 로컬로 컨플릭 해결한 뒤 force push with lease로 겨우겨우 살려서 돌아왔습니다 ㅠㅠ

derrickkim0109 replied:
고생하셨습니다..

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/Cells+SupplementaryViews/NoteDetailHeaderView.swift:
```diff
@@ -0,0 +1,106 @@
+
+import UIKit
+
+class NoteDetailHeaderView: UICollectionReusableView, Reusable {
```
final이 빠진것 같아요.

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/IncludeNoteButton.swift:
```diff
@@ -0,0 +1,117 @@
+
+import UIKit
+
+class IncludeNoteButton: UIButton {
```
final이 빠졌습니다.

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/NoteCommentsView.swift:
```diff
@@ -0,0 +1,212 @@
+import Combine
+import UIKit
+
+class NoteCommentsView: UIView {
```
여기도 final이 없어요.

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/NoteDetailView.swift:
```diff
@@ -0,0 +1,140 @@
+
+import Shared
+
+class NoteDetailView: UIView {
```
final이 없어요

derrickkim0109 commented on Projects/Feature/Main/Interface/Sources/Views/WriteCommentView.swift:
```diff
@@ -0,0 +1,242 @@
+import Combine
+import UIKit
+
+class WriteCommentView: UIView, UITextViewDelegate {
```
final이 없어요.

derrickkim0109 commented on Projects/Feature/Main/Testing/Sources/MockGetNoteWithCommentsUseCase.swift:
```diff
@@ -0,0 +1,117 @@
+    public init() { }
+    
+    public func execute(noteID: Int) -> AnyPublisher<NoteComments, NoteError> {
+        return Just(
```
굿~!

derrickkim0109 reviewed (commented):
정말 수고 많으셨습니다~!

조금 단순한 부분들 코멘트 남겼는데 답변 부탁드리겠습니다.

감사합니다.

inwoodev committed:
refactor: 피드백을 반영한 수정사항

- noteID 파라미터 이름 통일
- 상속받지 않을 클래스 객체들에 final 키워드 추가
- success를 프로퍼티로 갖고 있는 dto들을 FeelinSuccessResponse모델로 통일

그외:
- conflict시 수월한 작업을 위해 FeelinAPI case들 줄넘김 처리

derrickkim0109 reviewed (approved):
수고하셨습니다~!
